network/src/connectivity_manager/mod.rs

```diff
@@ -459,15 +463,43 @@ where
             min(
                 conn_limit
                     .saturating_sub(outbound_connections.saturating_add(self.dial_queue.len())),
-                to_connect.len(),
+                count,
             )
         } else {
-            to_connect.len()
+            count
         };
-        to_connect
-            .choose_multiple(&mut self.rng, to_connect_size)
-            .map(|(peer_id, peer)| (**peer_id, (*peer).clone()))
-            .collect()
+
+        let network_id = self.network_context.network_id();
+        let role = self.network_context.role();
+        let peer_roles = if network_id.is_validator_network() {
+            network_id.p2p_roles(&role)
+        } else {
+            network_id.upstream_roles(&role)
+        };
+
+        let mut to_dial = Vec::new();
+        let mut connected_left = to_connect_size;
+        for peer_role in peer_roles {
+            if connected_left == 0 {
+                break;
+            }
+
+            if let Some(peers) = prioritized.get(peer_role) {
+                if connected_left >= peers.len() {
+                    peers
+                        .iter()
+                        .for_each(|(peer_id, peer)| to_dial.push((*peer_id, (*peer).clone())));
+                } else {
+                    peers
+                        .choose_multiple(&mut self.rng, connected_left)
+                        .for_each(|(peer_id, peer)| to_dial.push((*peer_id, (*peer).clone())));
+                }
+
+                connected_left = connected_left.saturating_sub(peers.len());
+            }
+        }
+
+        to_dial
```

Comment:
maybe this would be cleaner if we just did like

```
let peer_roles = /* .. */;
let mut eligible = (/* .. */).filter(|p| peer_roles.contains(p.1.role));
eligible.sort_by(|a, b| a.1.role.cmp(b.1.role));

if let Some(conn_limit) = /* .. */ {
    eligible.truncate(conn_limit);
}

eligible
```

and reorder the `PeerRole`s so they're in priority order (looks like you just need to move Upstream and Downstream above VFN)

Reply:
The problem here is situational comparison.

There are some times when you want to order the roles in one direction vs others (upstream vs downstream, validator vs fullnode).  This causes a bunch of confusion, and basically I need to create several comparators for it, which wasn't clear when I first worked through this.  So, I can't simply reorder the `PeerRole` so that they're in priority order, but instead can provide a comparison function that takes in extra info.

Reply:
`eligible.sort_by(|a, b| my_cool_cmp_fn(peer_roles, a.1.role, b.1.role))`

: p

Reply:
I was going to say I'm going to do that, but that's why it ended up the way it was :P